### PR TITLE
Fix SnapGene Parser

### PR DESCRIPTION
**Pull request by rodoyle**
It works, but it isn't pretty. There is one reproducible bug with a new type of segment we haven't seen before.

### CODE REVIEW

**Comment by rodoyle** [comment]
@bmcorser This lacks elegance, but I'd like to get something in front of the team for review earlier rather than later. The bulk of the code in the snapgene.__init__ should probably be put in something like snapgene.adapter.

File: README.md [hunk 1]
```diff
@@ -2,6 +2,12 @@ Desktop Genetics Parsing Library
 ================================
 
 So many file formats, so little time
+## Installation
+```bash
+$ git clone https://github.com/deskgen/dgparse
+$ cd dgparse
+$ sudo python setup.py
```
Comment (bmcorser) [inline README.md]: `error: no commands supplied`

File: dgparse/sequtils.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    """get the complement of this sequence G --> C"""
+    #Sequence will always be DNA in this context
+    if isinstance(seq_str, str):
+        trans_table = string.maketrans(b'ACGTacgt', b'TGCAtgca')
+        return seq_str.translate(trans_table)
+    else:
+        return seq_str.translate(unicode_table)
+
+
+def reverse(seq_str):
```
Comment (bmcorser) [inline dgparse/sequtils.py]: Should probably be `get_reverse` for consistency

File: dgparse/sequtils.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    def __getitem__(self, n):
+        if n not in self.keep:
+            return None
+        return unichr(n)
+
+    def __call__(self, s):
+        return unicode(s).translate(self)
+
+
+makefilter = Keeper
```
Comment (bmcorser) [inline dgparse/sequtils.py]: The `Keeper` class doesn't have a docstring and doesn't appear to be used in this repo

File: dgparse/sequtils.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+
+def translate(seq, table):
```
Comment (bmcorser) [inline dgparse/sequtils.py]: `get_translation`?

**Comment by bmcorser** [comment]
A bunch of classes and functions provided here in the `sequtils` modules aren't used in this repo. They should be removed.

File: dgparse/snapgene/main.py [hunk 5]
```diff
@@ -162,21 +159,24 @@ def snapgene(f):
...
+        except UnicodeEncodeError:
+            # intercept these
+            raise
         except:
             # this is not totally robust: the error is raised because of a
             # key error following the wrong number of bytes in the previous
             # segment. It is possible that this still gives a valid key,
             # despite earlier error.
-            raise Exception("Badly formed segment or missing segment. Current segment: %s Previous Segment: %s" % (seg, lastSeg))
+            raise FormatException("Badly formed segment or missing segment. Current segment: %s Previous Segment: %s" % (seg, lastSeg))
```
Comment (bmcorser) [inline dgparse/snapgene/main.py]: I prefer something like:
``` python
fmt_str = ('Badly formed segment or missing segment. '
           'Current segment: {0} Previous Segment: {1}')
FormatException(fmt_str.format(seg, lastSeg))
```
Something similar actually is used below.

**Comment by bmcorser** [comment]
Code [like this](https://github.com/DeskGen/dgparse/blob/riley/fix-snapgene/dgparse/snapgene/segments.py#L172-L178) is a strong candidate for refactoring to use a validation library, such as [Voluptuous](https://github.com/alecthomas/voluptuous) which is used in Manifest.

**Comment by bmcorser** [comment]
The tests fail if they are run from the repo root (because of lines [like this](https://github.com/DeskGen/dgparse/blob/riley/fix-snapgene/tests/test_snapgene.py#L42)), consider writing in a function to discover fixtures relative to current module ie.
``` python
def get_fixture(name):
    module_directory, _ = os.path.split(__file__)
    return os.path.join(module_directory, name)
```

**Comment by bmcorser** [comment]
Yes, agree there should probably an adapter that calls this code, but actual SnapGene parser code should continue to live in `dgparse`

**Comment by rodoyle** [comment]
@bmcorser I've finished implementing your comments, though some in a slightly different way. Would you kindly merge this so Jay doesn't need to remember to download this specific branch to get a working version of the code.

**Comment by bmcorser** [comment]
@rodoyle you can merge if you need to, but I haven't had a chance to review the `schema` module properly.

You could push this out to Jay via PyPI from this branch (and he could install with Pip) so I can carry on reviewing ...

**Pull request merged**